### PR TITLE
Avoid asm library to be initialized twice

### PR DESCRIPTION
**Pull request by ke-yu**
### Purpose

This PR is to fix defect [MAGN-7927 AccessViolationException is thrown out when shutting down asm library](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7927#) and [MAGN-7915 LibG throws exception when it is shutdown in Extension loader](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7915#)

The root cause is asm library is initialized twice. More details:

`ProtoFFI.ExtensionAppLoader` will monitor app domain's `AssemblyLoad` and `AssemblyResolve` event, and for any loaded assembly, if the assembly implements `IExtensionApplication` interface, call `StartUp()` method on the `IExtensionApplication` instance (and call `ShutDown()` if the process exits). 

As it is not a singleton (`FFIExectionManager` has a `ExtensionAppLoader` member), we finally have multiple `ExtensionAppLoader` instances, will cause asm library is initialized twice when LibG is loaded. 
This PR makes it as a singleton.

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@aparajit-pratap

### CODE REVIEW

**Comment by aparajit-pratap** [comment]
@ke-yu could you verify if your fix also fixes this bug: http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-7915

File: src/Engine/ProtoCore/FFI/FFIExecutionManager.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@ private FFIExecutionManager()
         }
 
         Dictionary<RuntimeCore, FFIExecutionSession> mSessions = new Dictionary<RuntimeCore, FFIExecutionSession>();
-        ExtensionAppLoader mApploader = new ExtensionAppLoader();
+        ExtensionAppLoader mApploader = ExtensionAppLoader.Instance;
```
Comment (sharadkjaiswal) [inline src/Engine/ProtoCore/FFI/FFIExecutionManager.cs]: FFIEXecutionManager is singleton, so how are we creating multiple instances of ExtensionAppLoader?

Comment (ke-yu) [inline src/Engine/ProtoCore/FFI/FFIExecutionManager.cs]: No, although it is singleton, but it isn't used like a singleton. When core is disposed, `FFIExecutionManager` will set `mSelf` to null and re-create one later on.

**Comment by ke-yu** [comment]
@aparajit-pratap thanks! Verified that defect get fixed as well. I've updated PR to include that defect.

**Comment by aparajit-pratap** [comment]
Minor point but here is a good link for implementing thread-safe singleton patterns: http://csharpindepth.com/articles/general/singleton.aspx

File: src/Engine/ProtoCore/FFI/FFIExecutionManager.cs (hunk 1 above)
Comment (sharadkjaiswal) [inline src/Engine/ProtoCore/FFI/FFIExecutionManager.cs]: Got it.

**Comment by ke-yu** [comment]
@aparajit-pratap , cool! Replace with `Lazy<T>`.

**Comment by aparajit-pratap** [comment]
LGTM!

**Pull request merged**